### ISSUE TITLE
RFC: validation findings on tutorials/ — duplicate-header check cannot fire (silent target leakage), plus 4 parity gaps

### ISSUE THREAD

**Issue by kurtvalcorza** [kurtvalcorza/mitra-classifier-pipeline#16]
## Scope

Independent validation pass on this repo's `tutorials/` notebooks, requested by the maintainer. Filed as an
RFC: these are findings and proposed directions for comment, not patches. Nothing was merged or edited, and
this is provenance of a review, not a sign-off.

Reviewed copies are **byte-identical to `main` @ `bab857e50a8f`** (git blob SHAs compared against the tree), so
every line reference below applies to HEAD as of filing.

**Depth: static review only.** No `autogluon.tabular[mitra]` runtime and no Mitra weights were available in
this pass, so nothing in these two notebooks was executed. Findings below are source reading, with one
runtime mechanism proof where noted. Each is stated so that a single execution would confirm or refute it.

Evidence levels used below: **executed** (observed at runtime), **inspected** (read from notebook source),
**mechanism-proven** (the notebook's own code path was not run, but the behaviour it depends on was
reproduced at runtime in isolation).

## Summary

Four findings, one of which I would treat as the priority for this repo: the duplicate-header check on the
tutorial's primary bring-your-own-data path cannot fire, and the failure mode is silent target leakage rather
than an error. The other three are parity gaps against `mitra-regressor-pipeline`, which already implements
each of them correctly — so in every case there is a reference implementation in the family to copy from.

To be explicit about the limits of this pass: **nothing here was executed.** Finding 1 includes a runtime
mechanism proof, but not of the notebook's own code path.

### 1. Duplicate-header rejection cannot fire on the upload path — silent target leakage (high)

*Expected:* an uploaded labelled CSV with a repeated column name is rejected, as the notebook's own error
message promises: `duplicate column names are not supported.`

*Observed:* the upload reader at `tutorials/mitra_classifier_colab.ipynb` (code line ~211 of the
concatenated code cells) is a bare `pd.read_csv(io.BytesIO(payload))`, and the only duplicate test runs
afterwards inside `prepare()` (~line 258):

```python
if df.columns.duplicated().any():
    raise ValueError(f'{name}: duplicate column names are not supported.')
```

By that point pandas has already de-duplicated the header by renaming the second occurrence, so the test can
never be true for CSV input. Reproduced at runtime in isolation: a header ending `..., s6, target, bmi` is
read back as `..., s6, target, bmi.1`, and `df.columns.duplicated().any()` is `False`.

*Consequence:* a duplicated feature silently enters training as two columns. More seriously, because

```python
FEATURE_COLUMNS = [c for c in train_data.columns if c != TARGET_COLUMN]   # ~line 353
```

a duplicated **target** column arrives as `target.1` and becomes a feature — direct target leakage, which
will present as an implausibly good holdout metric rather than as an error. That is the failure mode I would
most want a tutorial to make impossible, since the user has no signal that anything went wrong.

*Evidence:* inspected (notebook source) + mechanism-proven (pandas de-duplication and the resulting
`duplicated()` outcome reproduced at runtime).

*For comment / why this looks like an oversight rather than a choice:* this same notebook already does it
correctly on the inference path (~line 532), with a comment naming the exact hazard —
`# Check original names before pandas can rename duplicate headers.` — and the regressor sibling factors the
same logic into `read_csv_payload()`. The fix looks like reusing the helper this notebook already contains.

### 2. Inference companion has no artifact integrity gate before unpickling (medium-high)

*Expected:* parity with the rest of the family, which verifies an uploaded artifact archive against a
maintainer-supplied digest before deserializing it.

*Observed:* `EXPECTED_ZIP_SHA256` appears 4x in each of the other three inference notebooks in this family and
**0x** in `tutorials/mitra_classifier_predictor_inference_colab.ipynb`. That notebook computes and prints the
ZIP digest for the user's records only, has no verification branch, and — unlike its regressor counterpart —
carries no trust-boundary note in the prose, even though `TabularPredictor.load` deserializes Python objects
out of the uploaded archive. The notebook does not *claim* to verify, so this is a gap rather than an
inaccurate statement; it is, however, the one place in the family where an untrusted archive is unpickled
with neither an integrity check nor a warning.

*Evidence:* inspected.

*For comment:* port the `EXPECTED_ZIP_SHA256` parameter, its verify-or-abort branch, and the trust-boundary
markdown from the regressor companion. Filing this publicly because it is tutorial-scope hardening on a
maintainer-supplied bundle, not a live-service vulnerability — say the word if you would rather track it
privately.

### 3. No pretrained-vs-fine-tuned selection, and no minimum-holdout guard (medium)

*Expected:* family-consistent behaviour — compare the fine-tuned candidate against the pretrained baseline on
the holdout, decline to select on a holdout too small to distinguish them, and warn if the selected model
degraded.

*Observed:* this notebook has no selection block at all. When fine-tuning runs, the fine-tuned predictor is
what gets evaluated, exported and recorded, even where it is worse than the pretrained baseline; there is no
`MIN_SELECTION_HOLDOUT_ROWS` equivalent. The regressor sibling compares, guards on holdout size, warns on
degradation, and records the basis; both TabICLv2 tutorials do the same. The markdown here does not promise
auto-selection, so this is a robustness and family-consistency gap rather than a documentation defect.

*Evidence:* inspected (diff against the fully-read regressor plus targeted greps into this notebook).

### 4. Companion is missing two guards its regressor counterpart has (low-medium)

*Expected:* parity with the regressor companion, which raises if
`predictor.problem_type != 'regression'` and if the inference CSV already contains a `prediction` column.

*Observed:* this companion only *displays* `problem_type` (~line 165), so loading a regressor archive into it
is not rejected at load — it would fail later inside `predict_proba` with an error that does not point at the
cause. It also assigns `result['prediction']` and `result['probability_<class>']` (~lines 266-269) with no
check for pre-existing columns of those names, so a user column called `prediction` is silently overwritten
in `predictions.csv`.

*Evidence:* inspected.

### 5. `AUTOGLOUON_VERSION` misspelling (cosmetic)

Misspelled consistently in both notebooks in this repo, so it is functional; the regressor side spells it
`AUTOGLUON_VERSION`. Worth fixing only because it is user-visible in the printed environment summary, and
because the inconsistency will bite anyone grepping across the four repos.

*Evidence:* inspected.

## Concerns checked and cleared

Reporting these alongside the defects so the review is not only a defect list:

- **Rerun safety:** the stale-variable clear list was suspected incomplete on this side; checked, it is not.
- **Import hygiene / execution-order integrity:** pyflakes over the concatenated code cells found **no
  undefined names** in either notebook. The only diagnostics are benign re-imports (4 in the tutorial) and one
  unused `os` in the companion — and the re-imports arguably help, since they make cells independently
  runnable after a partial run.
- **Metadata:** the export records the pretrained-vs-fine-tuned mode, and `tutorial_run_metadata.json` is
  richer on run provenance than the TabICLv2 equivalent — the acquisition source and metrics that the
  TabICLv2 manifests omit are present here.
- **Duplicate labelled rows:** `prepare()` does count exact duplicate rows and warn about leakage, which is
  the right instinct; finding 1 is that the *column*-level check next to it is the one that cannot fire.

## Coverage gaps (unexecuted checks, not passes)

- **Everything above is static.** No `autogluon.tabular[mitra]` runtime and no Mitra weights were available,
  so the HF download path, the offline-resolver assertion, fine-tuning, export, the reload smoke test, and the
  companion end-to-end are all unrun. Findings 1, 3, 4 and 5 would each be confirmed or refuted by one
  execution; finding 2 is a source-level absence and needs no run.
- **Two of this repo's notebooks exceeded my reader's size limit** and were reviewed via a diff against the
  fully-read regressor plus targeted greps that quote this notebook's own lines, rather than a linear read.
  Line references are approximate (concatenated-code-cell numbering) — treat them as locators, not citations.
- **Not assessed:** the `>500 features` and class-count refusal paths, imbalance edges, and the CI validator's
  interaction with any of the above.

For contrast on what an executed pass adds: the two TabICLv2 notebooks were run end-to-end here and their
integrity gates were attacked with four deliberate negative tests (wrong pinned digest, wrong archive digest,
one tampered byte inside a self-consistent bundle, duplicate header) — all four aborted correctly. I would
expect this repo to come out comparably well once a Mitra runtime is available; that pass has not happened.

**Comment by kurtvalcorza** [comment]
Companion RFCs from the same validation pass, since several findings reference behaviour in a sibling repo:

- TabICLv2 classifier: https://github.com/kurtvalcorza/tabicl-classifier-pipeline/issues/10
- TabICLv2 regressor: https://github.com/kurtvalcorza/tabicl-regressor-pipeline/issues/10
- Mitra classifier: https://github.com/kurtvalcorza/mitra-classifier-pipeline/issues/16
- Mitra regressor: https://github.com/kurtvalcorza/mitra-regressor-pipeline/issues/14

Depth differed by repo: the two TabICLv2 tutorials were executed end-to-end (DIMER ZIP path only) and their integrity gates attacked with deliberate negative tests; the two Mitra repos are static review only, with no Mitra runtime available in this pass.
